qbft/decided.go

```diff
@@ -0,0 +1,99 @@
+package qbft
+
+import (
+	"fmt"
+	"github.com/bloxapp/ssv-spec/types"
+	"github.com/pkg/errors"
+)
+
+// UponDecided returns decided msg if decided, nil otherwise
+func (c *Controller) UponDecided(msg *SignedMessage) (*SignedMessage, error) {
+	// decided msgs for past (already decided) instances will not decide again, just return
+	if msg.Message.Height < c.Height {
+		return nil, nil
+	}
+
+	if err := validateDecided(
+		c.config,
+		msg,
+		c.Share,
+	); err != nil {
+		return nil, errors.Wrap(err, "invalid decided msg")
+	}
+
+	// get decided value
+	data, err := msg.Message.GetCommitData()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get decided data")
+	}
+
+	// did previously decide?
+	inst := c.InstanceForHeight(msg.Message.Height)
+	prevDecided := inst != nil && inst.State.Decided
+
+	// Mark current instance decided
+	if inst := c.InstanceForHeight(c.Height); inst != nil && !inst.State.Decided {
+		inst.State.Decided = true
+		if msg.Message.Round > inst.State.Round {
+			inst.State.Round = msg.Message.Round
```

Comment:
I think you might have a bug here. You're updating the consensus instance `inst`'s `Round` to be the same as the message's `Round` value, but the message can be for a height higher than that of `inst`.